plugins/out_s3/s3.c

```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
             flb_plg_error(ctx->ins, "'s3_key_format' must start with a '/'");
             return -1;
         }
+        if (strstr((char *) tmp, "$INDEX")) {
+            ctx->key_fmt_has_seq_index = FLB_TRUE;
+
+            /* Initialize local storage to store metadata files */
+            ret = s3_store_init(ctx);
+            if (ret == -1) {
+                flb_plg_error(ctx->ins, "Failed to initialize S3 storage: %s",
+                              ctx->store_dir);
+                return -1;
+            }
+
+            fs_stream = flb_fstore_stream_create(ctx->fs, "metadata");
+            if (!fs_stream) {
+                flb_plg_error(ctx->ins, "could not initialize metadata stream");
+                flb_fstore_destroy(ctx->fs);
+                ctx->fs = NULL;
+                return -1;
+            }
+
+            /* Construct directories and file path names */
+            tmp = fs_stream->path;
+            ctx->metadata_dir = flb_sds_create(tmp);
+            tmp = "/index_metadata";
+            ctx->metadata_dir = flb_sds_cat(ctx->metadata_dir, tmp, strlen(tmp));
+
+            ctx->seq_index_file = flb_sds_create(ctx->metadata_dir);
+            tmp = "/seq_index_";
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
+            tmp = (char *) &ctx->ins->name;
+            ctx->seq_index_file = flb_sds_cat(ctx->seq_index_file, tmp, strlen(tmp));
```

Comment:
Why concatenate the plugin name to the file name?

Reply:
The plugin name should usually be something like `s3.0` or `s3.1`. However a user can give it their own name using an alias. I have no clue what the allowed char-set is for aliases. Are you certain that whatever this name is- it can be a file name? I wonder if the string needs to be sanitized?

Reply:
The purpose of concatenating is to prevent a scenario where multiple different S3 output plugins that are initialized in the same FluentBit configuration file have to share one metadata file.

For the plugin name, that is a good point. Sanitizing the name seems like a lot of unnecessary work, so I will just concatenate the id instead, and that should be good enough.